esphome/components/internal_temperature/internal_temperature.cpp

```diff
@@ -0,0 +1,65 @@
+#include "internal_temperature.h"
+#include "esphome/core/log.h"
+#include "esphome/core/hal.h"
+
+#ifdef USE_ESP32
+#ifdef CONFIG_IDF_TARGET_ESP32
+#ifdef __cplusplus
+extern "C" {
+#endif
+uint8_t temprature_sens_read();
+#ifdef __cplusplus
+}
+#endif
+#else  // CONFIG_IDF_TARGET_ESP32
+#include "driver/temp_sensor.h"
+#endif  // CONFIG_IDF_TARGET_ESP32
+#endif  // USE_ESP32
+
```

Comment:
No need for the `#ifdef __cplusplus` as this is a `.cpp` file and will thus always compile as C++, and we prefer our own `USE_ESP32_VARIANT_*` macros over those from the IDF (also change this below please). I also added a comment to note why we define the prototype here manually.
```suggestion
#ifdef USE_ESP32
#ifdef USE_ESP32_VARIANT_ESP32
// there is no official API available on the original ESP32
extern "C" {
uint8_t temprature_sens_read();
}
#else  // USE_ESP32_VARIANT_ESP32
#include "driver/temp_sensor.h"
#endif  // USE_ESP32_VARIANT_ESP32
#endif  // USE_ESP32
```